client/src/partials/budget/analysis/analysis_budget.js

```diff
@@ -0,0 +1,436 @@
+      });
+      $scope.total.surplus = precision.round(totalSurplus);
+      $scope.total.deficit = precision.round(totalDeficit);  
+    }
+
+    function start(models) {
+      angular.extend($scope, models);
+      $scope.accounts.data.forEach(function (acct) {
+        if ((acct.type !== 'title') && (acct.balance === null)) {
+          acct.balance = 0.0;
+        }
+      });
+
+      filterAccounts($scope.accounts.data)
+      .then(addBudgetData);
+      parseAccountDepth($scope.accounts.data, $scope.accounts);
+      session.mode = 'budget-analysis';
+    }
+
+    function displayAccounts() {
+      dependencies.accounts.query = '/InExAccounts/' +enterprise_id;
+      // Process period
+      var periodCriteria;
+      var selectedPeriod = session.periods.filter(function (p) {
+        return p.id === config.period_id;
+      })[0];
+      if (selectedPeriod.period_number === 0) {
+        dependencies.budgets.query.where = ['period.fiscal_year_id=' + config.fiscal_year_id];
+      } else {
+        dependencies.budgets.query.where = ['period.fiscal_year_id=' + config.fiscal_year_id, 'AND', 'period.id='+selectedPeriod.id];
+      }
+      validate.refresh(dependencies, ['accounts', 'budgets'])
+      .then(start);
+    }
+
+    function previousFYBudget(selectedPreviousFY) {
+      // Get budget data for all previous fiscal years
+      $scope.fiscalYearBudget = {};
+      for(var fy in selectedPreviousFY) {
+        var budget = { 
+          id           : selectedPreviousFY[fy].id, 
+          data         : null, 
+          totalBudget  : 0, 
+          totalBalance : 0 
+        };
+        $scope.fiscalYearBudget[budget.id] = {};
+        getBudget(budget.id);
+      }
+
+      function getBudget(fiscal_year_id) {
+        // Work with a copy of dependencies.budgets structure
+        // for a temporary previous fiscal years budgets
+        dependencies.previousFYBudget = dependencies.budgets;
+        // Process period
+        var periodCriteria;
+        var selectedPeriod = session.periods.filter(function (p) {
+          return p.id === config.period_id;
+        })[0];
+        if (selectedPeriod.period_number === 0) {
+          dependencies.previousFYBudget.query.where = ['period.fiscal_year_id=' + fiscal_year_id];
+        } else {
+          dependencies.previousFYBudget.query.where = ['period.fiscal_year_id=' + fiscal_year_id, 'AND', 'period.period_number='+selectedPeriod.period_number];
+        }
+        validate.refresh(dependencies, ['previousFYBudget'])
+        .then(function (model) {
+          var data = model.previousFYBudget.data;
+          handleFiscalYearBudget(fiscal_year_id, data);
+        });
+      }
+
+      function handleFiscalYearBudget(fiscal_year_id, data) {
+        $scope.fiscalYearBudget[fiscal_year_id] = data;
+      }
+    }
+
+    function loadFiscalYears(models) {
+      angular.extend($scope, models);
+    }
+
+    appstate.register('enterprise', function (enterprise) {
```

Comment:
You do not need to use appstate here.  We provide a new service, called SessionService, that can give you the current enterprise, user, and project without needing a callback.
```js
angular.module('bhima.controlllers')
.controller('ExampleController', ['$scope', 'SessionService', BudgetAnalysisController]);

function BudgetAnalysisController($scope, SessionService) {
  $scope.enterprise = SessionService.enterprise;  // $scope.enterprise.id, $scope.enterprise.currency_id, etc are defined!
  $scope.user = SessionService.user;   // $scope.user.id, $scope.user.username, etc.. are defined!
  $scope.project = SessionSerivice.project;  // $scope.project.id, $scope.project.abbr, etc ... are defined!
}
```
This should help clean your code up.